### PR TITLE
Refactor API server emphasis on logging

### LINKED ISSUE #10053: print api request and response information to the console when log-level=debug
**Is this a BUG REPORT or FEATURE REQUEST? (leave only one on its own line)**

/kind feature

**Description**

In order to make troubleshooting easier it would be nice if podman would print api request and response information to the console when log-level=debug.  This should include include all headers including tokens, request methods, request uri, payloads, etc.  There should be enough to reconstruct the api calls using curl.  the 'oc' command from OpenShift does this very nicely when you add lots of -vvvvvvv's.

Since some of the payloads can be large, more granular options for verbosity could be implemented.  It's not so useful to get a tar file printed to my console, but I definitely want to be able to see the json of a manifest write operation.

**Steps to reproduce the issue:**

1. do something with podman that would initiate communication with a registry endpoint

2. add log-level=debug to cli

3. observe output and wish that it included api request and response information

**Have you tested with the latest version of Podman and have you checked the Podman Troubleshooting Guide?**

podman version 3.0.1

**Additional environment details (AWS, VirtualBox, physical, etc.):**

fedora 33

### PR DESCRIPTION
**Pull request by jwhonce**
* To aid in debugging log API request and response bodies at trace
  level. Events can be correlated using the X-Reference-Id.
* Server now echos X-Reference-Id from client if set, otherwise
  generates an unique id.
* Move logic for X-Reference-Id into middleware
* Change uses of Header.Add() to Set() when setting Content-Type
* Log API operations in Apache format using gorilla middleware
* Port server code to use BaseContext and ConnContext

Fixes #10053

Signed-off-by: Jhon Honce <jhonce@redhat.com>

### CODE REVIEW

File: go.mod [hunk 1]
```diff
@@ -67,7 +68,10 @@ require (
 	golang.org/x/crypto v0.0.0-20210513164829-c07d793c2f9a
 	golang.org/x/sync v0.0.0-20210220032951-036812b2e83c
 	golang.org/x/sys v0.0.0-20210906170528-6f6e22806c34
+	golang.org/x/tools v0.1.2
 	gopkg.in/yaml.v3 v3.0.0-20210107192922-496545a6307b
 	k8s.io/api v0.22.1
 	k8s.io/apimachinery v0.22.1
 )
+
+replace github.com/vbauerster/mpb/v7 => github.com/mtrmac/mpb/v7 v7.0.5-0.20210831125917-6bcc64f93d02
```
Comment (Luap99) [inline go.mod:77]: Remove this line, also you are missing a `make vendor`

File: pkg/api/server/server.go [hunk 2]
```diff
@@ -205,20 +223,20 @@ func (s *APIServer) Serve() error {
 
 	go func() {
 		<-s.idleTracker.Done()
-		logrus.Debugf("API Server idle for %s", s.idleTracker.Duration.Round(time.Second).String())
+		logrus.Debug("API service shutting down idle for " + s.idleTracker.Duration.Round(time.Second).String())
```
Comment (mheon) [inline pkg/api/server/server.go:226]: Maybe a colon between "shutting down" and "idle"? Or reword to "shutting down after idling for"?

**Comment by TomSweeneyRedHat** [comment]
LGTM
and happy green test buttons.

**Pull request merged**